langchain4j-mongodb-atlas/src/main/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStore.java

```
@@ -342,13 +379,43 @@ private void createCollection(MongoDatabase database, String collectionName, Cre
         database.createCollection(collectionName, createCollectionOptions);
     }
 
-    private boolean isIndexExist(String indexName) {
+    private boolean indexExists(String indexName) {
+        return indexExistsAndNot(indexName, "DOES_NOT_EXIST");
+    }
+
+    private boolean indexExistsAndNot(String indexName, String... status) {
         return StreamSupport.stream(collection.listSearchIndexes().spliterator(), false)
-                .anyMatch(index -> indexName.equals(index.getString("name")));
+                .anyMatch(index -> {
+                    boolean sameName = indexName.equals(index.getString("name"));
+                    boolean hasStatus = Arrays.asList(status).contains(index.getString("status"));
+                    return sameName && !hasStatus;
+                });
     }
 
     private void createIndex(String indexName, IndexMapping indexMapping) {
-        Document index = fromIndexMapping(indexMapping);
-        collection.createSearchIndex(indexName, index);
+        collection.createSearchIndexes(Arrays.asList(new SearchIndexModel(
+                indexName,
+                fromIndexMapping(indexMapping),
+                SearchIndexType.vectorSearch())));
+
+        waitForIndex(indexName);
+    }
+
+
+    private void waitForIndex(String indexName) {
+        long startTime = System.nanoTime();
+        long timeoutNanos = TimeUnit.SECONDS.toNanos(SECONDS_TO_WAIT_FOR_INDEX);
+        while (System.nanoTime() - startTime < timeoutNanos) {
+            if (indexExistsAndNot(indexName, "INITIAL_SYNC", "PENDING", "BUILDING")) {
```

Index also might end up being in `FAILED` status which they can enter due to an invalid index definition. Should we throw an exception in this case if we have detected a failure within a given waiting time period?

---

I have added this, and also used the `queryable` flag to check if we should proceed (instead of the list of statuses).